### PR TITLE
Improve SmallEnumSet

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: In anticipation of using this to represent sets of CacheEntryRole for including or excluding kinds of blocks in block cache tiers, add significant new features to SmallEnumSet, including at least:

* List initialization
* Applicative constexpr operations
* copy/move/equality ops
* begin/end/const_iterator for iteration
* Better comments

Test Plan: unit tests added/expanded

### CODE REVIEW

File: include/rocksdb/data_structure.h [hunk 1]
```diff
@@ -15,37 +15,183 @@
...
-    uint64_t tmp = 1;
-    state_ |= (tmp << value);
-    return old_state != state_;
+  // copy
+  SmallEnumSet(const SmallEnumSet& that) = default;
+  SmallEnumSet& operator=(const SmallEnumSet& that) = default;
+
+  // move
+  SmallEnumSet(SmallEnumSet&& other) noexcept = default;
+  SmallEnumSet& operator=(SmallEnumSet&& other) noexcept = default;
```
Comment (ltamasi) [inline include/rocksdb/data_structure.h:60]: To follow the rule of zero, we could consider removing all these and let the compiler generate them. (The same goes below for the iterator.)

File: include/rocksdb/data_structure.h [hunk 2]
```diff
@@ -15,37 +15,183 @@
...
+    const_iterator(const_iterator&& that) noexcept = default;
+    const_iterator& operator=(const_iterator&& that) noexcept = default;
+
+    // equality
+    bool operator==(const const_iterator& that) const {
+      assert(set_ == that.set_);
+      return this->pos_ == that.pos_;
+    }
+
+    bool operator!=(const const_iterator& other) const {
```
Comment (ltamasi) [inline include/rocksdb/data_structure.h:97]: How about `that` instead of `other` for consistency?

File: util/slice_test.cc [hunk 3]
```diff
@@ -173,13 +173,69 @@ class SmallEnumSetTest : public testing::Test {
...
+
+TEST_F(SmallEnumSetTest, SmallEnumSetTest2) {
+  MyEnumClassSet s;  // based on an enum class type
+  ASSERT_TRUE(s.Add(MyEnumClass::A));
+  ASSERT_TRUE(s.Contains(MyEnumClass::A));
+  ASSERT_FALSE(s.Contains(MyEnumClass::B));
+  ASSERT_TRUE(s.With(MyEnumClass::B).Contains(MyEnumClass::B));
+  ASSERT_FALSE(s.Contains(MyEnumClass::B));
+  ASSERT_FALSE(s.Without(MyEnumClass::A).Contains(MyEnumClass::A));
+  ASSERT_TRUE(s.Contains(MyEnumClass::A));
```
Comment (ltamasi) [inline util/slice_test.cc:207]: Would be nice to test mixing `With` and `Without` here

**Review by ltamasi — APPROVED** [review]
LGTM in general. One thing I'm wondering about is if we could have leveraged `std::bitset` for the low-level bit manipulation part.

File: include/rocksdb/data_structure.h [hunk 4]
```diff
@@ -15,37 +15,183 @@
...
+}  // namespace detail
+
+// Represents a set of values of some enum type with a small number of
+// possible enumerators. For now, it supports enums where no enumerator
+// exceeds 63 when converted to int.
+template <typename ENUM_TYPE, ENUM_TYPE MAX_ENUMERATOR>
 class SmallEnumSet {
+ private:
+  using StateT = uint64_t;
+  static constexpr int kStateBits = sizeof(StateT) * 8;
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply include/rocksdb/data_structure.h:29]: For better or worse, there are many places where we assume char is 8 bits and zero existing places we rely on CHAR_BIT.

**Comment by pdillinger** [comment]
> One thing I'm wondering about is if we could have leveraged std::bitset for the low-level bit manipulation part.

Good point, though I sort of worry about unintended creep in the size of these sets if items are added to existing enums or sets are created over other larger enums. Right now it's small enough to use pass-by-value semantics. Thoughtful consideration should be given to expansion beyond 64 bits.

**Pull request closed without merging**